app/src/main/java/com/ubergeek42/WeechatAndroid/fragments/BufferFragment.java

```diff
@@ -351,14 +360,124 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
     ////////////////////////////////////////////////////////////////////////////////////////////////
 
     private boolean tcInProgress;
+    private boolean tcViaRelayInProgress = false;
+
+    @MainThread @SuppressLint("SetTextI18n") private void tryTabComplete() {
+        if (buffer == null) return;
+        // https://github.com/weechat/weechat/issues/1484
+        // Since WeeChat 2.9, relay clients can use the WeeChat completion mechanism.
+
+        // check for tcInProgress here if the previous tab complete ran
+        if (RelayConnection.weechatVersion >= 0x2090000 && !tcInProgress) {
+            tabCompleteRelayStrategy();
+        } else {
+            tabCompleteLocalStrategy();
+        }
+    }
+
+    private String tcViaRelayBaseString; // the entire original input
+    private int tcViaRelayIndex; // index used when looping over tcViaRelayCompletionResults
+    private String[] tcViaRelayCompletionResults;
+    private boolean tcViaRelayAddSpace; // to add a space after completion
+    private int tcViaRelayPosStart; // beginning of section to be replaced by one of the suggestions
+    private int tcViaRelayPosEnd; // end of said section in tcViaRelayBaseString
+    // tcViaRelayBaseString.subString(tcViaRelayPosStart, tcViaRelayPosEnd + 1)
+    private String tcViaRelayBaseWord;
+    private int tcViaRelaySelectionIndex; // where the cursor was when tab was pressed
+
+    @MainThread private void tabCompleteRelayStrategy() {
+        // disable the uiInput, re-enable it after the future has been completed
+        uiInput.setFocusable(false);
```

Comment:
we might call `setFocusable(true)` right after; perhaps only set it to false here if we don't do that?

Reply:
>     * perhaps there's a way to make the input field read-only without making it not focusable? maybe allow editing and simply cancel the completion attempt if user tries to edit the text?

Fixed, this should do the trick, I tested in the simulator and on a physical AOSP 6.0.1 device; the keyboard doesn't get dismissed upon pressing tab & focus is restored with the blinking cursor:
https://github.com/ubergeek42/weechat-android/blob/5b0d1d8faf2c9acc1e105f2f2647ef9f4b9396e2/app/src/main/java/com/ubergeek42/WeechatAndroid/tabcomplete/TabCompleter.kt#L65-L72

Reply:
i still think we shouldn't be touching the input field... for whatever reason this screws up the keyboard; in gboard the big white button briefly disappears, send button changes to return, etc

same goes for disabling the tab button, this just looks ugly